### PR TITLE
Sdk refactoring

### PR DESCRIPTION
**Pull request by mwamedacen**
Check at [README](https://github.com/paraswap/paraswap-sdk/blob/sdk-refactoring/README.md) for full details

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,151 +1,177 @@
...
+<p align="center">
+  <a href="https://paraswap.io">
+    <img src="https://cdn.paraswap.io/brand/paraswap.png" width="400px" >
+  </a>
+</p>
 
-<img src="https://paraswap-achirecture.netlify.com/logo.png" width="400px" >
 
----
+# Accompanion SDK of the ParaSwap API
```
Comment (samthomson) [inline README.md:8]:
```suggestion
# Companion SDK for the ParaSwap API
```
or maybe less is even more...
```suggestion
# SDK for the ParaSwap API
```

File: README.md [hunk 2]
```diff
@@ -1,151 +1,177 @@
...
+Refer to the documentation of the ParaSwap API: https://developers.paraswap.network
 
-https://developers.paraswap.network
 
-### To use ParaSwap SDK :
+## Features
+**Versatility**: works with both [web3](https://www.npmjs.com/package/web3) and [ethers](https://www.npmjs.com/package/ethers) without direct dependency
 
-Install the lib using npm or yarn
+**Canonicity**: bring only the functions you actually need
```
Comment (samthomson) [inline README.md:16]: I would just say 
```suggestion
**Canonical**: bring only the functions you actually need
```

File: README.md [hunk 3]
```diff
@@ -1,151 +1,177 @@
...
+
+const paraswap = constructSDK({
+  network: 1,
+  fetcher,
+  contractCaller,
+});
 ```
 
-##### To retrieve the list all available tokens:
+### To approve ParaSwap contracts to swap a ERC20 token
```
Comment (samthomson) [inline README.md:52]:
```suggestion
### To approve ParaSwap contracts to swap an ERC20 token
```

File: README.md [hunk 4]
```diff
@@ -1,151 +1,177 @@
...
+);
 
-const allowance = await paraSwap.getAllowance(userAddress, tokenAddress);
+const transactionResponse = await signer.sendTransaction(txParams);
+const transactionReceipt = await transactionResponse.wait();
 ```
 
-##### To approve an ERC20
+## Playground
+Interact with the ParaSwap SDK in CodeSandbox playground [here](https://codesandbox.io/s/gallant-flower-7yuker?file=/src/App.tsx)
```
Comment (samthomson) [inline README.md]:
```suggestion
Interact with the ParaSwap SDK in a CodeSandbox playground [here](https://codesandbox.io/s/gallant-flower-7yuker?file=/src/App.tsx)
```

File: README.md [hunk 5]
```diff
@@ -1,151 +1,177 @@
...
 
-```javascript
-const paraSwap = new ParaSwap().setWeb3Provider(web3Provider);
+## Bundle Optimization
+For bundle-size savvy developers, you can construct a lightweight version of this sdk to bring only the functions you need.
```
Comment (samthomson) [inline README.md:119]:
```suggestion
For bundle-size savvy developers, you can construct a lightweight version of the SDK and bring only the functions you need.
```

File: README.md [hunk 6]
```diff
@@ -1,151 +1,177 @@
...
-https://developers.paraswap.network
 
-### To use ParaSwap SDK :
+## Features
+**Versatility**: works with both [web3](https://www.npmjs.com/package/web3) and [ethers](https://www.npmjs.com/package/ethers) without direct dependency
 
-Install the lib using npm or yarn
+**Canonicity**: bring only the functions you actually need
+
+**Lightweight**: 400B Gzipped for the most minimalistic variant
```
Comment (samthomson) [inline README.md:18]:
```suggestion
**Lightweight**: 400B Gzipped for the minimal variant
```

File: README.md [hunk 7]
```diff
@@ -1,151 +1,177 @@
...
 ```
 
-##### To get the rate of a token pair using the API:
+### To get the rate of a token pair
 
-```javascript
+```typescript
 const srcToken = '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'; // ETH
-const destToken = '0x89d24a6b4ccb1b6faa2625fe562bdd9a23260359'; // DAI
+const destToken = '0xcAfE001067cDEF266AfB7Eb5A286dCFD277f3dE5'; // PSP
```
Comment (samthomson) [inline README.md:62]: :muscle:

File: README.md [hunk 8]
```diff
@@ -1,151 +1,177 @@
...
+Interact with the ParaSwap SDK in CodeSandbox playground [here](https://codesandbox.io/s/gallant-flower-7yuker?file=/src/App.tsx)
 
-```javascript
-const paraSwap = new ParaSwap().setWeb3Provider(web3Provider);
+## Bundle Optimization
+For bundle-size savvy developers, you can construct a lightweight version of this sdk to bring only the functions you need.
 
-const txHash = await paraSwap.approveToken(amount, userAddress, tokenAddress);
-```
+For e.g. for fetching rate and allowance only
```
Comment (samthomson) [inline README.md:121]:
```suggestion
e.g. for only getting rates and allowances
```
"For example" and "e.g." are synonymous, so best to use only one.

File: README.md [hunk 9]
```diff
@@ -1,151 +1,177 @@
...
-);
+const priceRoute = await minParaSwap.getRate(params)
+const allowance = await minParaSwap.getAllowance(userAddress, tokenAddress);
 ```
 
----
-
-### To run the example locally:
+## Legacy
+The ParaSwap class is exposed with some effort on backward compatibility with previous versions.
```
Comment (samthomson) [inline README.md:138]:
```suggestion

The `ParaSwap` class is exposed for backwards compatibility with previous versions of the SDK.
```

File: README.md [hunk 10]
```diff
@@ -1,151 +1,177 @@
...
-```bash
-yarn install paraswap
 ```
 
-For local developement you can run
-
-```bash
-yarn dev
-```
+By analogy to ```constructPartialSDK```, you can leverage a lightweight version of the sdk for fetching only
```
Comment (samthomson) [inline README.md:160]:
```suggestion
Which corresponds to ```constructPartialSDK``` in the SDK's current version to leverage a lightweight version for fetching only.
```

**Review by samthomson — COMMENTED** [review]
Minor grammar tweaks :straight_ruler:

File: README.md (hunk 7 above)
Comment (mwamedacen) [reply README.md:62]: 🐂

File: README.md (hunk 10 above)
Comment (mwamedacen) [reply README.md]: That's not exactly what I meant here. In the example above, we show how to use the ParaSwap class like before (with web3, axios). In this second example we show we can use the ParaSwap with 0 heavy dependancy which is sort of analogous to ```constructPartialSDK``` (with less bundle weight overhead though)

File: src/legacy/index.ts [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+  fetcher: FetcherFunction;
+
+  constructor(
+    private network: number = 1,
+    private apiURL: string = API_URL,
+    public web3Provider?: Web3,
+    public ethersDeps?: EthersProviderDeps, // need to be a provider with signer for approve requests
+    public account?: Address,
+    public axios?: AxiosStatic,
+    public fetch?: Fetch
```
Comment (mwamedacen) [inline src/legacy/index.ts:58]: I'm reiterating a previous concern, I think we should go for object params

**Review by ghost — APPROVED** [review]
app

**Review by ghost — COMMENTED** [review]
- Approve all

**Pull request merged**